cmake/sca/eclair/sca.cmake

```diff
@@ -0,0 +1,143 @@
+endif()
+if (CONFIG_ECLAIR_SUMMARY_ODT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-summary_odt=${ECLAIR_OUTPUT_DIR}/summary_odt")
+endif()
+if (CONFIG_ECLAIR_FULL_TXT_ALL_AREAS)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,areas")
+endif()
+if (CONFIG_ECLAIR_FULL_TXT_FIRST_AREA)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,first_area")
+endif()
+if (CONFIG_ECLAIR_FULL_TXT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-full_txt=${ECLAIR_OUTPUT_DIR}/txt")
+endif()
+if (CONFIG_ECLAIR_FULL_DOC_ALL_AREAS)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,areas")
+endif()
+if (CONFIG_ECLAIR_FULL_DOC_FIRST_AREA)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-setq=report_areas,first_area")
+endif()
+if (CONFIG_ECLAIR_FULL_DOC)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-full_doc=${ECLAIR_OUTPUT_DIR}/doc")
+endif()
+if (CONFIG_ECLAIR_FULL_ODT)
+  set(ECLAIR_REPORT_ADDITIONAL_OPTIONS "${ECLAIR_REPORT_ADDITIONAL_OPTIONS}" "-full_odt=${ECLAIR_OUTPUT_DIR}/odt")
+endif()
+
+message(STATUS "ECLAIR outputs have been written to: ${ECLAIR_OUTPUT_DIR}")
+message(STATUS "ECLAIR ECB files have been written to: ${ECLAIR_DATA_DIR}")
+message(STATUS "ECLAIR BUILD DIR is: ${ECLAIR_BUILD_DIR}")
+
+add_custom_target(eclair ALL
+  COMMAND ${CMAKE_COMMAND} -E
+    remove_directory ${ECLAIR_OUTPUT_DIR}
+  COMMAND ${CMAKE_COMMAND} -E
+    make_directory ${ECLAIR_OUTPUT_DIR}
+  COMMAND ${CMAKE_COMMAND} -E
+    make_directory ${ECLAIR_DATA_DIR}
+  COMMAND ${CMAKE_COMMAND} -E env
+    ECLAIR_DIAGNOSTICS_OUTPUT=${ECLAIR_DIAGNOSTICS_OUTPUT}
+    ECLAIR_DATA_DIR=${ECLAIR_DATA_DIR}
+    CC_ALIASES=${CC_ALIASES}
+    CXX_ALIASES=${CXX_ALIASES}
+    AS_ALIASES=${AS_ALIASES}
+    LD_ALIASES=${LD_ALIASES}
+    AR_ALIASES=${AR_ALIASES}
+    ${ECLAIR_ENV}
+    -verbose
+    -project_name=${ECLAIR_PROJECT_NAME}
+    -project_root=${ZEPHYR_BASE}
+    -eval_file=${ECLAIR_ECL_DIR}/analysis.ecl
+    -eval_file=${ECLAIR_ANALYSIS_ECL_DIR}/analysis_${ECLAIR_RULESET}.ecl
+    ${ECLAIR_ENV_ADDITIONAL_OPTIONS}
+    -- west build -p always
+                  -b ${BOARD}${BOARD_QUALIFIERS}
+                  --build-dir ${ECLAIR_BUILD_DIR}
+                  ${APPLICATION_CONFIG_DIR}
+  COMMAND ${CMAKE_COMMAND} -E env
+    ECLAIR_DATA_DIR=${ECLAIR_DATA_DIR}
+    ECLAIR_OUTPUT_DIR=${ECLAIR_OUTPUT_DIR}
+    ECLAIR_PROJECT_ECD=${ECLAIR_PROJECT_ECD}
+    ${ECLAIR_REPORT}
+    -quiet
+    -eval_file=${ECLAIR_ECL_DIR}/db_generation.ecl
+  COMMAND ${ECLAIR_REPORT}
+    -db=${ECLAIR_PROJECT_ECD}
+    ${ECLAIR_REPORT_ADDITIONAL_OPTIONS}
+    -overall_txt=${ECLAIR_OUTPUT_DIR}/overall.txt
+  COMMAND ${CMAKE_COMMAND} -E
+    cat ${ECLAIR_OUTPUT_DIR}/overall.txt
+  VERBATIM
+  USES_TERMINAL
+  COMMAND_EXPAND_LISTS
+)
```

Comment:
This creates a new target, so additionally to the regular build, you invoke `west build` with the `eclair` environment?

This feels like a lot of overhead actually, preferably the build should be replaced somehow.

Reply:
Maybe this should work similar to how `sparse` works?

Reply:
Come to think of it, don't you have infinite recursion if `ZEPHYR_SCA_VARIANT` is set in the environment?

Reply:
this was also pointed out in this comment: https://github.com/zephyrproject-rtos/zephyr/pull/68324#discussion_r1658651566
but was at that time accepted because 1) eclair was marked experimental, 2) eclair was targetting 3.7, meaning less time to address issues.

As Eclair missed 3.7, then perhaps take a second look to see if it can be done in this PR.
At lease give some feedback regarding this part in the earlier comment:
> Please note that west is an optional (but extremely useful) tool, but that still means we generally should not rely on west: https://docs.zephyrproject.org/latest/develop/west/without-west.html#using-zephyr-without-west
>
> Can this be changed into a new CMake call, and then build can be done as ${CMAKE_COMMAND} --build <path> ?

Reply:
> this was also pointed out in this comment: [#68324 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/68324#discussion_r1658651566) but was at that time accepted because 1) eclair was marked experimental, 2) eclair was targetting 3.7, meaning less time to address issues.
> 
> As Eclair missed 3.7, then perhaps take a second look to see if it can be done in this PR. At lease give some feedback regarding this part in the earlier comment:
> 
> > Please note that west is an optional (but extremely useful) tool, but that still means we generally should not rely on west: [docs.zephyrproject.org/latest/develop/west/without-west.html#using-zephyr-without-west](https://docs.zephyrproject.org/latest/develop/west/without-west.html#using-zephyr-without-west)
> > Can this be changed into a new CMake call, and then build can be done as ${CMAKE_COMMAND} --build  ?

I try to resolve this problem right now and I'm looking at a way similar to what sparse does by replacing the actual build with the intercepted compiler by eclair which also executes the build itself. This would also reduce the overhead